### DISCUSSION TITLE
SHA384/512 gives wrong hash on linux-aarch64

### DISCUSSION THREAD

**Discussion by manjitduhan** [openssl/openssl#22651] [Q&A] [answered]
The hash calculated by the openssl using SHA384/512 is wrong on linux-aarch64 while on "linux-octeon32", "x86", "x86_64" it is correct.

Openssl version: Checked with Openssl version 1.0.1p / 1.0.2k / 1.0.2l / 1.1.0i
Command used: echo $(echo -n "hello" | ./openssl dgst -sha512)

Output:
linux-aarch64: - 92496a87ea2af6f3bf72cdfb26596b0c2d589232d714098ea8594d3ed5ad134b16d3c8f8a213157477e3c965d03274f143798024e1afa523de715051281ba280

linux-octeon32: 9b71d224bd62f3785d96d46ad3ea3d73319bfbc2890caadae2dff72519673ca72323c3d99ba5c11d7c7acc6e14b8c5da0c4663475c2e5c3adef46f73bcdec043
x86: 9b71d224bd62f3785d96d46ad3ea3d73319bfbc2890caadae2dff72519673ca72323c3d99ba5c11d7c7acc6e14b8c5da0c4663475c2e5c3adef46f73bcdec043
x86_64: 9b71d224bd62f3785d96d46ad3ea3d73319bfbc2890caadae2dff72519673ca72323c3d99ba5c11d7c7acc6e14b8c5da0c4663475c2e5c3adef46f73bcdec043

we can see that the output on all platform is same and correct except for linux-aarch64.

Configure Output on linux-aarch64:

CC= /<path>/aarch64-2019.02/bin/aarch64-linux-gnu-gcc

CFLAG= -fPIC -DOPENSSL_PIC -DOPENSSL_THREADS  -DDSO_DLFCN -DHAVE_DLFCN_H -g -DOPENSSL_NO_HEARTBEATS -nostdinc -Wall -fpic -fPIC -Werror -Wunused -Wno-unused-but-set-variable -Wno-format -Wno-write-strings -Wno-parentheses -Wno-address -Wno-narrowing  -Wno-error -Wno-narrowing -g -I/<path>/aarch64-2019.02/lib/gcc/aarch64-linux-gnu/7.4.1/include -I/<path>/aarch64-2019.02/lib/gcc/aarch64-linux-gnu/x.x.x/include-fixed -Wno-error -DOS_LINUX -DARCH_ARM -D__LINUX_ARM_ARCH__=8 -DAP010 -D_UNIX -DNDEBUG -DBUILD_SYSTEM_VERSION=2 -DDEBUG -DCV_DEBUG  -isystem/<path>/aarch64-2019.02/aarch64-linux-gnu/libc/usr/include/ -O3 -DB_ENDIAN -DTERMIO  -Wall

DEPFLAG= -DOPENSSL_NO_EC_NISTP_64_GCC_128 -DOPENSSL_NO_GMP -DOPENSSL_NO_JPAKE -DOPENSSL_NO_MD2 -DOPENSSL_NO_RC5 -DOPENSSL_NO_RFC3779 -DOPENSSL_NO_SCTP -DOPENSSL_NO_STORE -DOPENSSL_NO_UNIT_TEST

PEX_LIBS= 
EX_LIBS= -ldl
EXE_EXT= 
ARFLAGS= 
AR= /<path>/aarch64-2019.02/bin/aarch64-linux-gnu-ar $(ARFLAGS) r
RANLIB= /<path>/aarch64-2019.02/bin/aarch64-linux-gnu-ranlib
NM= nm
PERL= /usr/bin/perl
TAR= tar
TARFLAGS= --no-recursion --record-size=10240
MAKEDEPPROG=makedepend
LIBDIR=lib

All other functionality is working fine except the wrong HASH calculation using SHA384/SHA512.

Is this know issue or any fix available in any build?

**Comment by mattcaswell** [comment]
> Checked with Openssl version 1.0.1p / 1.0.2k / 1.0.2l / 1.1.0i

These are all very old and out of support versions. Please can you check again with a currently supported version of OpenSSL (e.g. 3.1.4 or 3.0.12)

**Comment by manjitduhan** [comment]
Thanks @mattcaswell for very quick reply.

Sure, I will try on 3.x but need some time as there is huge change inbetween 3.x & where we are. 

By any chance, have you (or anyone else) come across this kind of issue i.e. SHA384/512 have failed on any processor?

**Comment by mattcaswell** [comment]
I don't recall seeing this before - but that's no guarantee that we haven't!

**Comment by jamuir** [comment]
I don't think any of the hashes posted are correct.

Here is what I get:

```
$ echo -n "hello" | sha256sum
2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824  -
$ echo -n "hello" | openssl dgst -sha256
SHA2-256(stdin)= 2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824
```

do you have `sha256sum` on those platforms?

**Comment by jamuir** [comment]
I see the problem.  The command posted in the description has `-sha256`, but the outputs are actually for `-sha512`:

```
$ echo -n "hello" | sha512sum 
9b71d224bd62f3785d96d46ad3ea3d73319bfbc2890caadae2dff72519673ca72323c3d99ba5c11d7c7acc6e14b8c5da0c4663475c2e5c3adef46f73bcdec043  -
$ echo -n "hello" | openssl dgst -sha512
SHA2-512(stdin)= 9b71d224bd62f3785d96d46ad3ea3d73319bfbc2890caadae2dff72519673ca72323c3d99ba5c11d7c7acc6e14b8c5da0c4663475c2e5c3adef46f73bcdec043
```
I agree the `9b71...` hashes are correct.

**Comment by manjitduhan** [comment]
@jamuir : yes, it was a typo. I have corrected it to SHA512.

**Comment by tom-cosgrove-arm** [comment]
@MannuSD So can I double-check: there is no problem, and this issue can be closed?

**Comment by manjitduhan** [comment]
> @MannuSD So can I double-check: there is no problem, and this issue can be closed?

@tom-cosgrove-arm : Problem is still there. On  linux-aarch64 based system the SHA512 & SHA384 is not correctly computed. I can re-confirm that this issue is on multiple versions of openssl 1.0.1, 1.0.2 and 1.1.0.

Regarding the above comment: While describing the issue in description section SHA256 got typed instead of SHA512 (which I corrected). So, you can ignore that comment.

We have started checking on openssl 3.1 but it will take time to adopt the changes in our system.

So, I would like to request to not close the issue for now...

**Comment by tom-cosgrove-arm** [comment]
> this issue is on multiple versions of openssl 1.0.1, 1.0.2 and 1.1.0

Note that these versions of OpenSSL are all end of life, and so won't get any fixes.

If there are issues with 3.x we will absolutely fix those

**Comment by jamuir** [comment]
> On linux-aarch64 based system the SHA512 & SHA384 is not correctly computed. I can re-confirm that this issue is on multiple versions of openssl 1.0.1, 1.0.2 and 1.1.0.

when I use OpenSSL 1.1.1 on aarch64 the correct hash is computed:

```
$ openssl version
OpenSSL 1.1.1d  10 Sep 2019
pi@raspberrypi:~ $ echo -n "hello" | openssl dgst -sha512
(stdin)= 9b71d224bd62f3785d96d46ad3ea3d73319bfbc2890caadae2dff72519673ca72323c3d99ba5c11d7c7acc6e14b8c5da0c4663475c2e5c3adef46f73bcdec043
pi@raspberrypi:~ $ echo -n "hello" | sha512sum 
9b71d224bd62f3785d96d46ad3ea3d73319bfbc2890caadae2dff72519673ca72323c3d99ba5c11d7c7acc6e14b8c5da0c4663475c2e5c3adef46f73bcdec043  -
```

**Comment by manjitduhan** [comment]
> > On linux-aarch64 based system the SHA512 & SHA384 is not correctly computed. I can re-confirm that this issue is on multiple versions of openssl 1.0.1, 1.0.2 and 1.1.0.
> 
> when I use OpenSSL 1.1.1 on aarch64 the correct hash is computed:
> 
> ```
> $ openssl version
> OpenSSL 1.1.1d  10 Sep 2019
> pi@raspberrypi:~ $ echo -n "hello" | openssl dgst -sha512
> (stdin)= 9b71d224bd62f3785d96d46ad3ea3d73319bfbc2890caadae2dff72519673ca72323c3d99ba5c11d7c7acc6e14b8c5da0c4663475c2e5c3adef46f73bcdec043
> pi@raspberrypi:~ $ echo -n "hello" | sha512sum 
> 9b71d224bd62f3785d96d46ad3ea3d73319bfbc2890caadae2dff72519673ca72323c3d99ba5c11d7c7acc6e14b8c5da0c4663475c2e5c3adef46f73bcdec043  -
> ```

@jamuir : Thank you for trying it at your end and confirming that it is working fine.

Will it be possible for you to share the Configure output? I just want to make sure that we are not missing any flag or misconfigured it at our end.

**Comment by jamuir** [comment]
I didn't compile that version myself.  It is included with Raspberry Pi OS.

Here is the output of `openssl version -a`, which may be helpful:

```
$ openssl version -a
OpenSSL 1.1.1d  10 Sep 2019
built on: Tue Aug 24 08:30:43 2021 UTC
platform: debian-arm64
options:  bn(64,64) rc4(char) des(int) blowfish(ptr) 
compiler: gcc -fPIC -pthread -Wa,--noexecstack -Wall -Wa,--noexecstack -g -O2 -fdebug-prefix-map=/build/openssl-RzZXhH/openssl-1.1.1d=. -fstack-protector-strong -Wformat -Werror=format-security -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_BN_ASM_MONT -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DVPAES_ASM -DECP_NISTZ256_ASM -DPOLY1305_ASM -DNDEBUG -Wdate-time -D_FORTIFY_SOURCE=2
OPENSSLDIR: "/usr/lib/ssl"
ENGINESDIR: "/usr/lib/aarch64-linux-gnu/engines-1.1"
Seeding source: os-specific
```

**Comment by jamuir** [comment]
Just built 1.1.1 myself and the hash is computed correctly.

```
$ cd ~/tmp
$ wget https://www.openssl.org/source/openssl-1.1.1w.tar.gz
$ tar xfz openssl-1.1.1w.tar.gz
$ cd openssl-1.1.1w/
$ echo ./Configure linux-aarch64 -Werror --strict-warnings --debug > doit.bash
$ echo make -j2 >> doit.bash
$ echo make test >> doit.bash
$ . doit.bash &> log.txt
```
The contents of `log.txt` show that all tests were successful.

```
$ LD_LIBRARY_PATH=. ./apps/openssl version
OpenSSL 1.1.1w  11 Sep 2023
$ echo -n "hello" | LD_LIBRARY_PATH=. ./apps/openssl dgst -sha512
(stdin)= 9b71d224bd62f3785d96d46ad3ea3d73319bfbc2890caadae2dff72519673ca72323c3d99ba5c11d7c7acc6e14b8c5da0c4663475c2e5c3adef46f73bcdec043
```

@MannuSD : can you try out the above steps and tell us if it works for you?  This could be done fairly quickly, and it probably makes sense to figure out the sha512 failure before investing a lot of time migrating to 3.x.

**Comment by manjitduhan** [comment]
> Just built 1.1.1 myself and the hash is computed correctly.
> 
> ```
> $ cd ~/tmp
> $ wget https://www.openssl.org/source/openssl-1.1.1w.tar.gz
> $ tar xfz openssl-1.1.1w.tar.gz
> $ cd openssl-1.1.1w/
> $ echo ./Configure linux-aarch64 -Werror --strict-warnings --debug > doit.bash
> $ echo make -j2 >> doit.bash
> $ echo make test >> doit.bash
> $ . doit.bash &> log.txt
> ```
> 
> The contents of `log.txt` show that all tests were successful.
> 
> ```
> $ LD_LIBRARY_PATH=. ./apps/openssl version
> OpenSSL 1.1.1w  11 Sep 2023
> $ echo -n "hello" | LD_LIBRARY_PATH=. ./apps/openssl dgst -sha512
> (stdin)= 9b71d224bd62f3785d96d46ad3ea3d73319bfbc2890caadae2dff72519673ca72323c3d99ba5c11d7c7acc6e14b8c5da0c4663475c2e5c3adef46f73bcdec043
> ```
> 
> @MannuSD : can you try out the above steps and tell us if it works for you? This could be done fairly quickly, and it probably makes sense to figure out the sha512 failure before investing a lot of time migrating to 3.x.

Thank you @jamuir . Definitely, we will take it as first thing tomorrow and will update the result here. 
Thanks once again for trying it proactively.

**Comment by manjitduhan** [comment, marked as answer]
We have tried to rebuild openssl 1.1.1d. The issue is not observed and it working fine.

 ./openssl version
OpenSSL 1.1.1d  10 Sep 2019 (Library: OpenSSL 1.1.1  11 Sep 2018)

echo $(echo -n "hello" | ./openssl dgst -sha512)
(stdin)= 9b71d224bd62f3785d96d46ad3ea3d73319bfbc2890caadae2dff72519673ca72323c3d99ba5c11d7c7acc6e14b8c5da0c4663475c2e5c3adef46f73bcdec043

Thanks @jamuir once again.
